src/HSE/All.hs

```diff
@@ -139,44 +157,80 @@ runCpp (Cpphs o) file x = dropLine <$> runCpphs o file x
+  { DynFlags.sTargetPlatform=platform
+  , DynFlags.sPlatformConstants=platformConstants
+  , DynFlags.sProjectVersion=cProjectVersion
+  , DynFlags.sProgramName="ghc"
+  , DynFlags.sOpt_P_fingerprint=fingerprint0
+  }
+  where
+    platform =
+      Platform{platformWordSize=8
+              , platformOS=OSUnknown
+              , platformUnregisterised=True}
+    platformConstants =
+      DynFlags.PlatformConstants{DynFlags.pc_DYNAMIC_BY_DEFAULT=False,DynFlags.pc_WORD_SIZE=8}
+
+fakeLlvmConfig :: (DynFlags.LlvmTargets, DynFlags.LlvmPasses)
+fakeLlvmConfig = ([], [])
+
 -- | Parse a Haskell module. Applies the C pre processor, and uses best-guess fixity resolution if there are ambiguities.
---   The filename @-@ is treated as @stdin@. Requires some flags (often 'defaultParseFlags'), the filename, and optionally the contents of that file.
-parseModuleEx :: ParseFlags -> FilePath -> Maybe String -> IO (Either ParseError (Module SrcSpanInfo, [Comment]))
+-- The filename @-@ is treated as @stdin@. Requires some flags (often 'defaultParseFlags'), the filename, and optionally the contents of that file.
+-- This version used both hs-src-exts AND ghc-lib. It's considered to be an unrecoverable error if one
+-- parsing method succeeds whilst the other fails.
+parseModuleEx :: ParseFlags -> FilePath -> Maybe String -> IO (Either ParseError ParsedModuleResults)
 parseModuleEx flags file str = timedIO "Parse" file $ do
         str <- case str of
             Just x -> return x
             Nothing | file == "-" -> getContentsUTF8
                     | otherwise -> readFileUTF8' file
         str <- return $ fromMaybe str $ stripPrefix "\65279" str -- remove the BOM if it exists, see #130
         ppstr <- runCpp (cppFlags flags) file str
-        case parseFileContentsWithComments (mode flags) ppstr of
-            ParseOk (x, cs) -> return $ Right (applyFixity fixity x, cs)
-            ParseFailed sl msg -> do
-                -- figure out the best line number to grab context from, by reparsing
-                -- but not generating {-# LINE #-} pragmas
+        case (parseFileContentsWithComments (mode flags) ppstr, parseFileGhcLib file ppstr)
+          of
+            (ParseOk (x, cs), Lexer.POk _ mod) ->
+              return $ Right (ParsedModuleResults (applyFixity fixity x, cs) mod)
+            (ParseFailed sl msg, Lexer.PFailed ps) -> do
                 flags <- return $ parseFlagsNoLocations flags
                 ppstr2 <- runCpp (cppFlags flags) file str
                 let pe = case parseFileContentsWithMode (mode flags) ppstr2 of
                         ParseFailed sl2 _ -> context (srcLine sl2) ppstr2
                         _ -> context (srcLine sl) ppstr
-                return $ Left $ ParseError sl msg pe
+                return $ Left $ ParseError sl msg pe (pprErrMsgBagWithLoc $ snd $ Lexer.getMessages ps dynFlags)
+            (_, _) -> error "Unexpected : one but not both parse methods failed."
```

Comment:
I expect this to be super common. Question is what to do about it...

Reply:
What has been determined is that there are no existing tests where GHC gives a parse and HSE fails. There are however 32 (of 656) tests where HSE gives a successful parse but GHC reports a failure. Arranged the code for the time being so that these tests continue to pass.